builtin/providers/aws/resource_aws_waf_rule.go

```diff
@@ -0,0 +1,202 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/waf"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsWafRule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsWafRuleCreate,
+		Read:   resourceAwsWafRuleRead,
+		Update: resourceAwsWafRuleUpdate,
+		Delete: resourceAwsWafRuleDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"metric_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"predicates": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"negated": &schema.Schema{
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"data_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								value := v.(string)
+								if len(value) > 128 {
+									errors = append(errors, fmt.Errorf(
+										"%q cannot be longer than 128 characters", k))
+								}
+								return
+							},
+						},
+						"type": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								value := v.(string)
+								if value != "IPMatch" && value != "ByteMatch" && value != "SqlInjectionMatch" && value != "SizeConstraint" && value != "XssMatch" {
+									errors = append(errors, fmt.Errorf(
+										"%q must be one of IPMatch | ByteMatch | SqlInjectionMatch | SizeConstraint | XssMatch", k))
+								}
+								return
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsWafRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	// ChangeToken
+	var ct *waf.GetChangeTokenInput
+	time.Sleep(10 * time.Second) // To Prevent WAFStaleDataException
+	res, err := conn.GetChangeToken(ct)
+	if err != nil {
+		return fmt.Errorf("Error getting change token: %s", err)
+	}
+
+	params := &waf.CreateRuleInput{
+		ChangeToken: res.ChangeToken,
+		MetricName:  aws.String(d.Get("metric_name").(string)),
+		Name:        aws.String(d.Get("name").(string)),
+	}
+
+	resp, err := conn.CreateRule(params)
+	if err != nil {
+		return err
+	}
+	d.SetId(*resp.Rule.RuleId)
```

Comment:
Is there a chance Rule can be nil? If so, we should guard against this being a dereferencing error